### PR TITLE
issue #303: Added test case for BulkCacheLoadingException

### PR DESCRIPTION
**Pull request by palmanojkumar**

### CODE REVIEW

File: core/src/test/java/org/ehcache/EhcacheBasicCrudBase.java [hunk 1]
```diff
@@ -730,6 +731,13 @@ public void describeTo(final Description description) {
      * Keys for which access results in a thrown {@code Exception}.  This set may be empty.
      */
     private final Set<String> failingKeys;
+    
+    /**
+     * Keys for which success should be reported in {@link BulkCacheLoadingException}
+     */
+    private Set<String> successKeys;
+    
+    private boolean isBulkCacheLoadingExceptionEnabled = false;
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicCrudBase.java]: Why do you need these two additional fields?
I was expecting you would be able to use the `org.ehcache.EhcacheBasicCrudBase.FakeCacheLoaderWriter#completeFailureKey` and `org.ehcache.EhcacheBasicCrudBase.FakeCacheLoaderWriter#failingKeys` as they are used in the `BulkCacheWritingException` testing

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 2]
```diff
@@ -2337,6 +3310,55 @@ public void testGetAllStoreAllMatchCacheAccessExceptionBeforeLoaderSomeMatchDisj
    * <ul>
    *    <li>non-empty request key set</li> 
    *    <li>all {@link Store} entries match</li>
+   *    <li>{@link Store#bulkComputeIfAbsent} throws before accessing loader</li>
+   *    <li>some {@link CacheLoaderWriter} entries match</li>
+   *    <li>non-matching {@link CacheLoaderWriter#loadAll(Iterable)} calls fail witb {@link BulkCacheLoadingexception}</li>
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Two typos on this line:
* witb -> with
* BulkCacheLoadingexception -> `BulkCacheLoadingException`

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 3]
```diff
@@ -366,6 +443,99 @@ public void testGetAllStoreNoMatchCacheAccessExceptionBeforeLoaderAllFail() thro
...
+    final InOrder ordered = inOrder(this.loaderWriter, this.spiedResilienceStrategy);
+    verify(this.store).bulkComputeIfAbsent(eq(KEY_SET_A), getAnyIterableFunction());
+    // ResilienceStrategy invoked: no assertion for Store content
+    ordered.verify(this.loaderWriter, atLeast(1)).loadAll(this.loadAllCaptor.capture());
+    assertThat(this.getLoadAllArgs(), equalTo(KEY_SET_A));
+
+    ordered.verify(this.spiedResilienceStrategy)
+        .getAllFailure(eq(KEY_SET_A), any(CacheAccessException.class), this.bulkExceptionCaptor.capture());
+    assertThat(this.bulkExceptionCaptor.getValue().getSuccesses().keySet(), empty());
+    assertThat(this.bulkExceptionCaptor.getValue().getFailures().keySet(), Matchers.<Set<?>>equalTo(KEY_SET_A));
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: You have some duplication caused by the fact that you check here the exception but also in the catch block above.
I would either check here that we have the same exception or extract that code to an exception check method:
`checkBulkLoadingException(BulkCacheLoadingException e, Set<K> successKeys, Set<K> failureKeys)`

Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Note that the comment above applies to all added test methods that use the `bulkExceptionCaptor`

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 4]
```diff
@@ -1306,15 +1828,99 @@ public void testGetAllStoreSomeMatchLoaderNoMatchSomeFail() throws Exception {
...
+   *    <li>no {@link CacheLoaderWriter} entries match</li>
+   *    <li>some {@link CacheLoaderWriter#loadAll(Iterable)} calls fail with {@link BulkCacheLoadingException}</li>
+   * </ul>
+   */
+  @Test
+  public void testGetAllStoreSomeMatchLoaderNoMatchSomeFailWithBulkCacheLoadingException() throws Exception {
+    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B));
+    this.store = spy(fakeStore);
+
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_F), KEY_SET_C, true);
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 5]
```diff
@@ -1306,15 +1828,99 @@ public void testGetAllStoreSomeMatchLoaderNoMatchSomeFail() throws Exception {
...
+   * </ul>
+   */
+  @Test
+  public void testGetAllStoreSomeMatchCacheAccessExceptionBeforeLoaderNoMatchSomeFail() throws Exception {
+    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B));
+    this.store = spy(fakeStore);
+    doThrow(new CacheAccessException("")).when(this.store)
+        .bulkComputeIfAbsent(getAnyStringSet(), getAnyIterableFunction());
+
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_F), KEY_SET_D);
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 6]
```diff
@@ -1327,27 +1933,29 @@ public void testGetAllStoreSomeMatchLoaderNoMatchSomeFail() throws Exception {
...
   @Test
-  public void testGetAllStoreSomeMatchCacheAccessExceptionBeforeLoaderNoMatchSomeFail() throws Exception {
+  public void testGetAllStoreSomeMatchCacheAccessExceptionBeforeLoaderNoMatchAllFailWithBulkCacheLoadingException() throws Exception {
     final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B));
     this.store = spy(fakeStore);
     doThrow(new CacheAccessException("")).when(this.store)
         .bulkComputeIfAbsent(getAnyStringSet(), getAnyIterableFunction());
 
-    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_F), KEY_SET_D);
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_F), true);
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 7]
```diff
@@ -1562,6 +2170,47 @@ public void testGetAllStoreSomeMatchLoaderSomeMatchDisjointFail() throws Excepti
...
+   *    <li>some {@link CacheLoaderWriter} entries match</li>
+   *    <li>non-matching {@link CacheLoaderWriter#loadAll(Iterable)} calls fails with {@link BulkCacheLoadingException}</li>
+   * </ul>
+   */
+  @Test
+  public void testGetAllStoreSomeMatchLoaderSomeMatchDisjointFailWithBulkCacheLoadingException() throws Exception {
+    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B));
+    this.store = spy(fakeStore);
+
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_E, KEY_SET_E), KEY_SET_C, true);
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.
Also you use twice the same key set here.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 8]
```diff
@@ -1607,6 +2256,55 @@ public void testGetAllStoreSomeMatchCacheAccessExceptionBeforeLoaderSomeMatchDis
...
+   * </ul>
+   */
+  @Test
+  public void testGetAllStoreSomeMatchCacheAccessExceptionBeforeLoaderSomeMatchDisjointFailWithBulkCacheLoadingException() throws Exception {
+    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B));
+    this.store = spy(fakeStore);
+    doThrow(new CacheAccessException("")).when(this.store)
+        .bulkComputeIfAbsent(getAnyStringSet(), getAnyIterableFunction());
+
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_E), KEY_SET_E, true);
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 9]
```diff
@@ -1740,28 +2480,167 @@ public void testGetAllStoreSomeMatchCacheAccessExceptionBeforeLoaderAllMatchNone
    * </ul>
    */
   @Test
-  public void testGetAllStoreSomeMatchCacheAccessExceptionAfterLoaderAllMatchNoneFail() throws Exception {
-    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B), Collections.singleton("keyA3"));
+  public void testGetAllStoreSomeMatchCacheAccessExceptionAfterLoaderAllMatchNoneFail() throws Exception {
+    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B), Collections.singleton("keyA3"));
+    this.store = spy(fakeStore);
+
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_A, KEY_SET_C, KEY_SET_D));
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 10]
```diff
@@ -1740,28 +2480,167 @@ public void testGetAllStoreSomeMatchCacheAccessExceptionBeforeLoaderAllMatchNone
...
+   * Tests {@link Ehcache#getAll(Set)} for
+   * <ul>
+   *    <li>non-empty request key set</li> 
+   *    <li>all {@link Store} entries match</li>
+   *    <li>{@link Store#bulkComputeIfAbsent} throws before accessing loader</li>
+   *    <li>no {@code CacheLoaderWriter}</li>
+   * </ul>
+   */
+  @Test
+  public void testGetAllStoreAllMatchCacheAccessExceptionBeforeNoLoader() throws Exception {
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Did you really add that test or is this diff having a hard time?

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 11]
```diff
@@ -2100,6 +3026,53 @@ public void testGetAllStoreAllMatchCacheAccessExceptionBeforeLoaderNoMatchSomeFa
...
+   * </ul>
+   */
+  @Test
+  public void testGetAllStoreAllMatchCacheAccessExceptionBeforeLoaderNoMatchSomeFailWithBulkCacheLoadingException() throws Exception {
+    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B));
+    this.store = spy(fakeStore);
+    doThrow(new CacheAccessException("")).when(this.store)
+        .bulkComputeIfAbsent(getAnyStringSet(), getAnyIterableFunction());
+
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_C, KEY_SET_F), KEY_SET_F, true);
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 12]
```diff
@@ -2374,6 +3396,51 @@ public void testGetAllStoreAllMatchCacheAccessExceptionAfterLoaderSomeMatchDisjo
...
+   *    <li>some {@link CacheLoaderWriter} entries match</li>
+   *    <li>non-matching {@link CacheLoaderWriter#loadAll(Iterable)} calls fail with {@link BulkCacheLoadingException}</li>
+   * </ul>
+   */
+  @Test
+  public void testGetAllStoreAllMatchCacheAccessExceptionAfterLoaderSomeMatchDisjointFailWithBulkCacheLoadingException() throws Exception {
+    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B), Collections.singleton("keyA3"));
+    this.store = spy(fakeStore);
+
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_B, KEY_SET_C), KEY_SET_B, true);
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java [hunk 13]
```diff
@@ -2337,6 +3310,55 @@ public void testGetAllStoreAllMatchCacheAccessExceptionBeforeLoaderSomeMatchDisj
...
+   * </ul>
+   */
+  @Test
+  public void testGetAllStoreAllMatchCacheAccessExceptionBeforeLoaderSomeMatchDisjointFailWithBulkCacheLoadingException() throws Exception {
+    final FakeStore fakeStore = new FakeStore(getEntryMap(KEY_SET_A, KEY_SET_B));
+    this.store = spy(fakeStore);
+    doThrow(new CacheAccessException("")).when(this.store)
+        .bulkComputeIfAbsent(getAnyStringSet(), getAnyIterableFunction());
+
+    final FakeCacheLoaderWriter fakeLoader = new FakeCacheLoaderWriter(getEntryMap(KEY_SET_B, KEY_SET_C), KEY_SET_B, true);
```
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Do not start in an inconsistent state: the loader needs to know a super set of what the store knows. If that makes the test a duplicate, delete it.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java (hunk 10 above)
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: Loaded changeset in IDE, this is diff having a hard time

File: core/src/test/java/org/ehcache/EhcacheBasicCrudBase.java (hunk 1 above)
Comment (palmanojkumar) [inline core/src/test/java/org/ehcache/EhcacheBasicCrudBase.java]: Now using failingKeys for BulkCacheLoadingException. isBulkCacheLoadingExceptionEnabled  flag is required to identify whether failingKeys are for general loading exception(ie RuntimeException) vs BulkCacheLoadingException

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java (hunk 2 above)
Comment (palmanojkumar) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: corrected it

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java (hunk 3 above)
Comment (palmanojkumar) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: yeah. I prefer extracting out code to checkBulkLoadingException method

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java (hunk 5 above)
Comment (palmanojkumar) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: fixed.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java (hunk 10 above)
Comment (palmanojkumar) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: There is problem with git diff. I havent modified existing test cases. sometime I feel svn diff is better than git diff

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java (hunk 4 above)
Comment (palmanojkumar) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: All remarks about inconsistence state is fixed.

File: core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java (hunk 9 above)
Comment (ljacomet) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: This has not been addressed

Comment (palmanojkumar) [inline core/src/test/java/org/ehcache/EhcacheBasicGetAllTest.java]: testGetAllStoreSomeMatchCacheAccessExceptionAfterLoaderAllMatchNoneFail() method is not changed at all. GIT diff is reporting it incorrectly.
However, such issues needs to be fixed and I propose opening a new issue for it.

**Pull request merged**